.github/workflows/windows.yml

```diff
@@ -48,40 +56,59 @@ env:
 jobs:
   build:
     name: Windows Build
-    runs-on: windows-2022
+    runs-on: ${{ inputs.self-hosted-runner && 'self-hosted-windows10' || 'windows-2022' }}
     steps:
-      - uses: actions/checkout@v4
-        if: github.event_name != 'pull_request_target'
-      # This is necessary to checkout the pull request if this run was triggered via a
-      # `pull_request_target` event.
-      - uses: actions/checkout@v4
-        if: github.event_name == 'pull_request_target'
+      - if: ${{ ! inputs.self-hosted-runner && github.event_name != 'pull_request_target' }}
+        uses: actions/checkout@v4
+      - if: ${{ ! inputs.self-hosted-runner && github.event_name == 'pull_request_target' }}
+        uses: actions/checkout@v4
         with:
           ref: ${{ github.event.pull_request.head.sha }}
           fetch-depth: 2
+
+      # Faster checkout for self-hosted runner that uses prebaked repo.
+      - if: ${{ inputs.self-hosted-runner && github.event_name != 'pull_request_target' }}
+        run: git fetch --depth=2 origin $env:GITHUB_SHA
+      - if: ${{ inputs.self-hosted-runner && github.event_name == 'pull_request_target' }}
+        run: git fetch --depth=2 origin refs/pull/${{ github.event_number }}/head
+      - if: ${{ inputs.self-hosted-runner }}
+        run: git switch --detach FETCH_HEAD
+
+      - name: ccache
+        # FIXME: “Error: Restoring cache failed: Error: Unable to locate executable file: sh.”
+        if: ${{ ! inputs.self-hosted-runner }}
+        uses: hendrikmuhs/ccache-action@v1.2
+
+      # Install missing tools in a GitHub-hosted runner.
       # Workaround for https://github.com/actions/runner-images/issues/10001:
       - name: Upgrade llvm
+        if: ${{ ! inputs.self-hosted-runner }}
         run: choco upgrade llvm
-      - name: ccache
-        uses: hendrikmuhs/ccache-action@v1.2
       - name: Install taplo
+        # Already installed in self-hosted runner (also, pwsh command not found).
+        if: ${{ ! inputs.self-hosted-runner }}
         uses: baptiste0928/cargo-install@v3
         with:
           crate: taplo-cli
           locked: true
       - name: Install cargo-deny
+        # Already installed in self-hosted runner (also, pwsh command not found).
+        if: ${{ ! inputs.self-hosted-runner }}
         uses: baptiste0928/cargo-install@v3
         with:
           crate: cargo-deny
           locked: true
       - name: Install wixtoolset
+        if: ${{ ! inputs.self-hosted-runner }}
         run: |
           choco install wixtoolset
           echo "C:\\Program Files (x86)\\WiX Toolset v3.11\\bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
       - uses: actions/setup-python@v5
+        if: ${{ ! inputs.self-hosted-runner }}
         with:
           python-version: "3.10"
       - name: Bootstrap
+        if: ${{ ! inputs.self-hosted-runner }}
```

Comment:
Question: How do we handle rust upgrades (usually needs now crown, and CI image will benefit of preinstalling newer rust)? We might need to add comment in rust-toolchain.toml?

Also how do we handle other changes to bootstrap? Maybe we should still run it, but make it such that it doesn't do extra work when not needed (if crown is already installed it will not be recompiled IIRC).

Reply:
rust upgrades are handled at runtime by mach, but not crown. baking a new image automatically updates rust, crown, and the other bootstrap tools.

in general, i’m undecided on whether we should run bootstrap at workflow runtime. i think it would be valuable to have the build fail loudly if the github-hosted environment diverges from the self-hosted environment, since it keeps the two environments aligned by requiring us to update the self-hosted environment when we land bootstrap changes.

Reply:
> baking a new image automatically updates rust, crown, and the other bootstrap tools

But this needs to be done manually, right? I think it would be nice to automate this using github workflows😁.

I am somehow worried, that diverging will be causing flaky hidden due to auto selection of runners.

Reply:
i agree, i plan to automate that too, when i have more time.

Reply:
My question is how to land a change that updates rustc (and crown). Let's say we have a upgrade PR. This PR cannot land because it needs new crown (rustc upgrade is done by rustup), if we build a new image then and deploy it, no PR can land in the mean time (it will cause disturbance and requires a lot of manual intervention) and even after it lands, all other PRs needs to be rebased for try runs (I think).

TL;DR I think we still need to run `cargo install --path support/crown` (it should be quick if crown is already installed and nothing changed).

Reply:
I think this still needs to be addressed.

Reply:
always installing crown seems like a good middle ground, since it mitigates the big rustc upgrade problem while having ~zero impact on build time and complexity. it might break incremental builds, but incremental builds are currently broken either way (looking into that for a later patch). done in 4887363267395196a3d8b1968eab25baafb2d523 ([build](https://github.com/servo/servo/actions/runs/10558475197)).